pkg/planner/core/operator/logicalop/logical_join.go

```diff
@@ -336,31 +336,47 @@ func (p *LogicalJoin) BuildKeyInfo(selfSchema *expression.Schema, childSchema []
 		if len(p.EqualConditions) == 0 {
 			return
 		}
-		lOk := false
-		rOk := false
-		// Such as 'select * from t1 join t2 where t1.a = t2.a and t1.b = t2.b'.
-		// If one sides (a, b) is a unique key, then the unique key information is remained.
-		// But we don't consider this situation currently.
-		// Only key made by one column is considered now.
-		evalCtx := p.SCtx().GetExprCtx().GetEvalCtx()
+		// Extract all left and right columns from equal conditions
+		leftCols := make([]*expression.Column, 0, len(p.EqualConditions))
+		rightCols := make([]*expression.Column, 0, len(p.EqualConditions))
 		for _, expr := range p.EqualConditions {
-			ln := expr.GetArgs()[0].(*expression.Column)
-			rn := expr.GetArgs()[1].(*expression.Column)
-			for _, key := range childSchema[0].PKOrUK {
-				if len(key) == 1 && key[0].Equal(evalCtx, ln) {
-					lOk = true
-					break
-				}
+			leftCols = append(leftCols, expr.GetArgs()[0].(*expression.Column))
+			rightCols = append(rightCols, expr.GetArgs()[1].(*expression.Column))
+		}
+		checkColumnsMatchPKOrUK := func(cols []*expression.Column, pkOrUK []expression.KeyInfo) bool {
+			if len(pkOrUK) == 0 {
+				return false
+			}
+			colSet := intset.NewFastIntSet()
+			for _, col := range cols {
+				colSet.Insert(int(col.UniqueID))
 			}
-			for _, key := range childSchema[1].PKOrUK {
-				if len(key) == 1 && key[0].Equal(evalCtx, rn) {
-					rOk = true
-					break
+			// Check if any key is a subset of the join key columns
+			// A match is considered valid if join key columns contain all columns of any PKOrUK
+			for _, key := range pkOrUK {
+				allMatch := true
+				for _, k := range key {
+					if !colSet.Has(int(k.UniqueID)) {
+						allMatch = false
+						break
+					}
+				}
+				if allMatch {
+					return true
 				}
 			}
+			return false
 		}
-		// For inner join, if one side of one equal condition is unique key,
-		// another side's unique key information will all be reserved.
+		lOk := checkColumnsMatchPKOrUK(leftCols, childSchema[0].PKOrUK)
+		rOk := checkColumnsMatchPKOrUK(rightCols, childSchema[1].PKOrUK)
+
+		// When both sides match different unique keys (e.g., left side matches unique key setA, right side matches unique key setB),
+		// we can derive additional functional dependencies through join conditions:
+		// 1. Original FDs: setA -> left row, setB -> right row
+		// 2. Join condition: setA from left = setA from right
+		// 3. By augmentation rule (X->Y implies XZ->YZ): setA from right + setB -> left row + setB
+		// 4. By transitivity rule (X->Y and Y->Z implies X->Z): setA from right + setB -> left row + right row
+		// This allows us to preserve unique key information from both sides when join keys provide stronger uniqueness guarantees.
 		// If it's an outer join, NULL value will fill some position, which will destroy the unique key information.
 		if lOk && p.JoinType != LeftOuterJoin {
```

Comment:
i think the below is missing
lOk + LeftOuter， we can add childSchema[0].PKOrUK...
rOk + RightOuter， we can add childSchema[1].PKOrUK...

Reply:
right duplicate key row will pollute the pkuk from left side, resolved